### PR TITLE
web3 interface api

### PR DESCRIPTION
**Pull request by DavidChild**
add web3 interface netversion,receipt

### CODE REVIEW

File: core/api/src/jsonrpc/impl.rs [hunk 1]
```diff
@@ -149,4 +151,50 @@ where
...
+        todo!();
+    }
+
+    async fn get_code(&self, _address: H160, _number: Option<U256>) -> RpcResult<Bytes> {
+        todo!();
+    }
+
+    async fn get_transaction_receipt(
+        &self,
+        _hash: H256,
```
Comment (KaoImin) [inline core/api/src/jsonrpc/impl.rs:178]: Why add a _ prefix?

File: core/api/Cargo.toml [hunk 2]
```diff
@@ -10,6 +10,7 @@ async-graphql = { version = "3.0", features = ["tracing"] }
 cita_trie = "3.0"
 jsonrpsee = { version = "0.6", features = ["http-server", "macros"] }
 parking_lot = "0.11"
+ethereum = "0.10.0"
```
Comment (KaoImin) [inline core/api/Cargo.toml:13]: Remove this, all the types used from crate ethereum are re-exported from protocol.

File: core/api/src/jsonrpc/impl.rs [hunk 3]
```diff
@@ -3,12 +3,13 @@ use crate::jsonrpc::{AxonJsonRpcServer, RpcResult};
 
 use jsonrpsee::types::Error;
 
+use crate::jsonrpc::types::{
```
Comment (KaoImin) [inline core/api/src/jsonrpc/impl.rs:6]: Our crate usage order is that: 
```rust
use std::xxx;

use outer_crate::xxx;

use common_crate::xxx;
use protocol::xxx;

use crate::xxx;
```

Move the `use` to the right position.

File: core/api/src/jsonrpc/impl.rs [hunk 4]
```diff
@@ -149,4 +151,50 @@ where
...
+        &self,
+        _hash: H256,
+    ) -> RpcResult<Option<Web3TransactionReceipt>> {
+        let tx = self
+            .adapter
+            .get_transaction_by_hash(Context::new(), _hash)
+            .await
+            .map_err(|e| Error::Custom(e.to_string()))?;
+
+        let rt = self
```
Comment (KaoImin) [inline core/api/src/jsonrpc/impl.rs:186]: The `rt` often means a runtime, change a name.

File: core/api/src/jsonrpc/impl.rs [hunk 5]
```diff
@@ -149,4 +151,50 @@ where
...
+            .adapter
+            .get_receipt_by_tx_hash(Context::new(), _hash)
+            .await
+            .map_err(|e| Error::Custom(e.to_string()))?;
+
+        match tx {
+            Some(y) => rt.map_or_else(
```
Comment (KaoImin) [inline core/api/src/jsonrpc/impl.rs:193]:
```suggestion
            Some(y) => {
                if let Some() = rt {
                    return;
                } else {
                    return;
                }
            }
```

File: core/api/src/jsonrpc/types.rs [hunk 6]
```diff
@@ -36,6 +37,49 @@ impl RichTransactionOrHash {
...
+    pub logs_bloom:          Bloom,
+    pub status:              U256,
+    pub to:                  Option<H160>,
+    pub transaction_hash:    Hash,
+    pub transaction_index:   Option<U256>,
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub transaction_type:    Option<U64>,
+}
+impl Web3TransactionReceipt {
+    pub fn from(receipt: Receipt, tx: SignedTransaction) -> Web3TransactionReceipt {
```
Comment (KaoImin) [inline core/api/src/jsonrpc/types.rs:59]: The `from` method is provided by std::convert::From, change a method name.

File: core/api/src/jsonrpc/types.rs [hunk 7]
```diff
@@ -36,6 +37,49 @@ impl RichTransactionOrHash {
...
+        Web3TransactionReceipt {
+            block_number:        receipt.block_number,
+            block_hash:          receipt.block_hash,
+            contract_address:    None,
```
Comment (KaoImin) [inline core/api/src/jsonrpc/types.rs:63]: Why is None here?

**Review by KaoImin — CHANGES_REQUESTED** [review]

File: core/api/src/jsonrpc/mod.rs [hunk 8]
```diff
@@ -46,6 +46,29 @@ pub trait AxonJsonRpc {
 
     #[method(name = "eth_estimateGas")]
     async fn estimate_gas(&self, req: CallRequest, number: Option<BlockId>) -> RpcResult<U256>;
+    #[method(name = "net_version")]
+    async fn net_work(&self) -> RpcResult<U256>;
```
Comment (KaoImin) [inline core/api/src/jsonrpc/mod.rs:50]:
```suggestion
    async fn net_version(&self) -> RpcResult<U256>;
```

File: core/api/src/jsonrpc/types.rs [hunk 9]
```diff
@@ -36,6 +37,49 @@ impl RichTransactionOrHash {
         }
     }
 }
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
```
Comment (KaoImin) [inline core/api/src/jsonrpc/types.rs:40]: Serde with camelCase

File: core/api/src/jsonrpc/types.rs [hunk 10]
```diff
@@ -36,6 +37,49 @@ impl RichTransactionOrHash {
         }
     }
 }
```
Comment (KaoImin) [inline core/api/src/jsonrpc/types.rs:39]: Add a new line below.

File: core/api/src/jsonrpc/types.rs [hunk 11]
```diff
@@ -36,6 +37,49 @@ impl RichTransactionOrHash {
...
+    pub gas_used:            U256,
+    pub logs:                Vec<Log>,
+    pub logs_bloom:          Bloom,
+    pub status:              U256,
+    pub to:                  Option<H160>,
+    pub transaction_hash:    Hash,
+    pub transaction_index:   Option<U256>,
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub transaction_type:    Option<U64>,
+}
```
Comment (KaoImin) [inline core/api/src/jsonrpc/types.rs:57]: Add a new line below.

**Review by KaoImin — CHANGES_REQUESTED** [review]

**Comment by KaoImin** [comment]
Change the PR name.

File: core/api/src/jsonrpc/types.rs (hunk 7 above)
Comment (DavidChild) [reply core/api/src/jsonrpc/types.rs:63]: If it is a contract transaction, then this address points to the new contract address. If it is an ordinary transaction, this field can be blank. At present, the underlying receipt does not return a constraint address, so it is temporarily replaced by none

**Pull request closed without merging**